doc/designs/fips_indicator.md

```diff
@@ -26,59 +26,122 @@ The following information was extracted from the FIPS 140-3 IG [5] “2.4.C Appr
+- The FIPS restrictions should be able to be enabled/disabled per algorithm context.
+- If the per algorithm context value is not set, then the  Configuration file option is used.
+
 Solution
 --------
 
-We already have most of the existing code in the FIPS provider using
+In OpenSSL most of the existing code in the FIPS provider is using
 implicit indicators i.e. An error occurs if existing FIPS rules are violated.
 
 The following rules will apply to any code that currently is not FIPS approved,
 but needs to be.
 
-- There will be a 'fips approved' mode per provider operation context that is on by default.
-- In this mode any FIPS rule that is violated will result in an error.
-- The 'fips approved' mode may be turned off per context via a OSSL_PARAM setter, if it
-is turned off then the operation is not FIPS approved.
-- A getter is required that returns if the operation is 'fips approved' using an
-OSSL_PARAM. The returned value can be a combination of the set 'fips_approved' mode
-plus any other logic. This is an explicit indicator.
-- Any algorithm that is transitioning [4] to not being allowed should be removed from
-the fips provider.
-- If an algorithm is transitioning [4] to be only allowed for processing
-(e.g. verification, signing, key validation) then the protection code
-(keygen, signing, encryption) should be removed from the fips provider, any
-attempt to use the protection API's should result in an error. The processing
-code should still be functional.
-
-Other rules:
-The existing flags that we set in the fips config file to control security checks
-will continue to function as they do now, and will not be affected by the
-strict mode variable.
-The getter to determine if we are fips approved will however take the flags
-into account.
-
-Motivation
+- The fipsinstall application will have a configurable item added for each algorithm that requires a change.
+These options will be passed to the FIPS provider in a manner similar to existing code.
+
+- A user defined callback similar to OSSL_SELF_TEST will be added. This callback
+will be triggered whenever an approved mode test fails.
+It may be set up by the user using
+
+```c
+OSSL_INDICATOR_set_callback(OSSL_LIB_CTX *libctx, OSSL_CALLBACK *cb, void *cbarg)
+```
+
+- The callback API will be
+
+```c
+int OSSL_INDICATOR_callback(OSSL_LIB_CTX *libctx, const char *algtype,
+                            const char *algdesc)
+```
+
+This can be used during application testing to log that an indicator was
+triggered. The user callback may return 0 if the application wants an error
+to occur based on the indicator type and description.
+
+- To control an algorithm context's checks via code requires a setter (e.g OSSL_ALG_PARAM_STRICT_CHECKS),
+
+```c
+    p = OSSL_PARAM_locate_const(params, OSSL_ALG_PARAM_STRICT_CHECKS);
+    if (p != NULL
+        && !OSSL_PARAM_get_int(p, &ctx->strict_checks))
```

Comment:
Nit: Is this backwards.  The text indicates controlling an algs context checks via code requires a setter, but the example code illustrates  a getter

Reply:
No its not backwards. This gets the value out of the p and sets it into ctx->strict_checks.
(Yes it is a bit confusing, but that is how the set_ctx code works)